[#4587] docs: Refactor the security document and add authorization push-down page (#4624)
apache/gravitino: docs/security/access-control.md

## Overview

Apache Gravitino(incubating) is a technical data catalog that uses a unified metadata paradigm to manage multiple data sources while still allowing multiple engines like Spark, Trino, and Flink, or Python to connect to these data sources for data processing through Gravitino.

Because each underlying data source will have its own access control system, it can be difficult to plug in data engines with the intent of querying multiple of these data at once.
This is especially important for data governance practitioners who have to worry about data access restrictions and data compliance issues, but this is streamlined through Gravitino.

> Therefore, in the hopes of solving this big data issue, Gravitino plans to implement a universal set of permissions models and paradigms.

Reviewer: ```suggestion
Therefore, in the hopes of solving this big data issue, Gravitino plans to implement a universal set of privilege models and paradigms.
```

With this, users will be able to manage all of their data sources on a single access plane, regardless of whether the data source is a database, or a message queue or an object storage system.

After authorizing these data sources within Gravitino’s metadata lake, authentication can then be performed in Spark, Trino, and Flink Engines, as well as our Python client.
This abstraction allows users to control access to data and make compliant use of the data without having to obstruct other teams and worrying about the tedious work of individual access control systems.

> ### Gravitino Permission Model

Reviewer: ```suggestion
### Gravitino Privilege Model
```

Gravitino’s unified management model allows for each data source to have its own authorization features. However, each data source may come with its own dedicated authorization model and methods.
We may not be able to properly set permissions to the underlying system, so when a given user tries to access this data, the underlying authorization system may result in permission inconsistencies and cause issues for external access.

> To mitigate this issue, Gravitino aims to provide a unified authorization model and accompanying methods that sit on top of all the data sources instead, making it much easier to manage access rights.

Reviewer: ```suggestion
To mitigate this issue, Gravitino aims to provide a unified authorization model and accompanying methods that sit on top of all the data sources instead, making it much easier to manage access privileges.
```

It is important to note that Gravitino’s authorization model will not merge the access control systems of the underlying data sources to form a large and unwieldy set of privileges.

> Instead, We will summarize the usage of the privileges currently in use within the data system, and offer a set of Gravitino-native permission models that accurately reflect it.

Reviewer: ```suggestion
Instead, We will summarize the usage of the privileges currently in use within the data system, and offer a set of Gravitino-native privilege models that accurately reflect it.
```

:::

## Concept

### Authorization

Gravitino also provides a set of authorization frameworks to interact with different underlying data source
authorization systems (e.g., MySQL's own access control management and the Apache Ranger access control management system for big data)
in accordance with its own authorization model and methodology.
More information you can see the [Authorization push down](authorization-pushdown.md).

### Authentication

> As mentioned above, Gravitino uses Ownership to control the rights of securable object in the management category and uses Role to control access securable objects,

Reviewer: ```suggestion
As mentioned above, Gravitino uses Ownership to control the privileges of securable object in the management category and uses Role to control access securable objects,
```

so when a user performs a specific operation on a specified resource,

> Gravitino will perform a composite authentication on the Ownership and Role to which the resource belongs.

Reviewer: ```suggestion
Gravitino will perform a composite authentication on the Ownership and Role to which the securable object belongs.
```

When a user has more than one Role, Gravitino will use the user's current Role for authentication, and the user can switch the current Role to access a different securable object.

### Role

> The traditional rights system generally uses RBAC (Role-Based Access Control) for rights management,

Reviewer: ```suggestion
The traditional access control system generally uses RBAC (Role-Based Access Control) for access control management,
```

where each Role contains a collection of different operating privileges for a different securable object.
When the system adds a new user or user group, you can select the Roles which they are expected to be granted to,

> so that the user can quickly start using it, without waiting for the administrator to gradually set up the access rights to securable object for him.

Reviewer: ```suggestion
so that the user can quickly start using it, without waiting for the administrator to gradually set up the access privileges to securable object for him.
```

then you need to design many different Privileges, however, too many Privileges will cause too complicated settings in the authorization.

If you only need to carry out coarse-grained privilege control on the securable object in the system, then you only need to design a small number of Privileges,
but it will result in too weak control ability when the authentication. Therefore, the design of Privilege is an important trade-off in the access control system.
We know that Privilege is generally divided into two types, one is the management category of Privilege, such as the `CREATE`, `DELETE` resource privilege,
and the other is the operation category of Privilege, such as the `READ` and `WRITE` resource privilege.

In most organizations, the number of data managers is much smaller than the number of data users.
Because it is the data users who need fine-grained privilege control,
we must provide more Privileges related to usage and more tightly gatekeeper the administrative Privileges.
To enforce this, we’ll introduce the concept of Ownership as a complete replacement for the administrative category of Privilege.

### Ownership

> When you create a resource (Gravitino Service, Metalake, Catalog, and any other entity) in Gravitino, each entity has an Owner field that defines the user (or group) to which the resource belongs.

Reviewer: ```suggestion
When you create a securable object (Gravitino Service, Metalake, Catalog, and any other entity) in Gravitino, each entity has an Owner field that defines the user (or group) to which the resource belongs.
```

> The owner of each entity has implicit administrative class privilege, for example, to delete that resource.

Reviewer: ```suggestion
The owner of each entity has implicit administrative class privilege, for example, to delete that securable object.
```

> Only the Owner of a resource can fully manage that resource.

Reviewer: ```suggestion
Only the Owner of a securable object can fully manage that resource.
```

> If a resource needs to be managed by more than one person at the same time, the owner is assigned to a user group.

Reviewer: ```suggestion
If a securable object needs to be managed by more than one person at the same time, the owner is assigned to a user group.
```

### User
Users are generally granted one or multiple Roles, and users have different operating privileges depending on their Role.

### Group

> To make it easier to grant a single access control to multiple users, we can add users to a user group, and then grant one or multiple Roles to that user group.

Reviewer: ```suggestion
To make it easier to grant a single access control to multiple users, we can add users to a user group, and then grant one or multiple roles to that user group.
```

> This process allows all users belonging to that user group to have the access control in those Roles.

Reviewer: ```suggestion
This process allows all users belonging to that user group to have the access control in those roles.
```

:::

## Concept

### Authorization

Gravitino also provides a set of authorization frameworks to interact with different underlying data source
authorization systems (e.g., MySQL's own permission management and the Apache Ranger permission management system for big data)
in accordance with its own authorization model and methodology.
More information you can see [Authorization push down](authorization-pushdown.md).

### Authentication

As mentioned above, Gravitino uses Ownership to control the rights of resources in the management category and uses Role to control the permissions of operations in the operation category,
so when a user performs a specific operation on a specified resource,
Gravitino will perform a composite authentication on the Ownership and Role to which the resource belongs.

> When a user has more than one Role, Gravitino will use the user's current Role for authentication, and the user can switch the current Role to access different resources.

Reviewer 2: This part seems to relate to authorization rather than authentication.

Author: The authorization relates to access control.
Here's just a brief description and add jump to authorization-pushdown.md

> #### Permission fuzzy matching

Reviewer: This is duplicated with object iheritance.

Gravitino uses the resource's ENTITY ID (long type) to preserve permission relationships.
So Gravitino can't directly support fuzzy matching of resource names (string type), such as wildcards like (*) and (%).
Gravitino uses the resource parent node to express support for all resource (*) wildcards for child resources, for example,
if we need to set read permissions for all table resources, we can set it to `{catalog1.schema1, READ_TABLE_PRIVILEGE}`,
which stands for having access to `catalog1.schema1.*` read access to all tables.

> ### Storage of authority data

Reviewer: This is the implement of user system.

Additionally, Gravitino provides a mechanism to ensure fault tolerance in the privilege data storage and push-down to the underlying data source privilege system.
As a result of this, Gravitino’s push-downs to the underlying permission system operations are required to be idempotent operation,
meaning that each kind of privilege push-down operation can be repeated without causing conflicts and will always achieve the same output.

### Role

The traditional rights system generally uses RBAC (Role-Based Access Control) for rights management,
where each Role contains a collection of different operating privileges for different resources.
When the system adds a new user or user group, you can select the Roles which they are expected to be granted to,
so that the user can quickly start using it, without waiting for the administrator to gradually set up the access rights to resources for him.

Roles also employ the concept of ownership – the owner of a Role is by default the creator of the Role,
implying the owner has all the permissions to operate the Role, including deleting the Role.

> Gravitino introduces the concept of the “Current” Role, where a user can own multiple Roles at the same time,

Reviewer: We don't have the concept of current role.
